man/vde_pcapplug.1
.TH VDE_PCAPPLUG 1 "August 20, 2008" "Virtual Distributed Ethernet"
.SH NAME
vde_pcapplug \- Virtual Distributed Ethernet libpcap cable
.SH SYNOPSIS
.B vde_pcapplug 
[OPTION]... \fIinterface\fP
.br
.SH DESCRIPTION
.B vde_pcapplug 
is a plug to be connected into a VDE switch.
All the data that is caught by the plug is injected into
.I interface 
and everything from that interface is sent in the switch.

.SH OPTIONS
.TP
\fB\-p\fP, \fB\-\-port\fP=\fIportnum\fP
It is possible to decide which port of the switch to use.
When this option is not specified the switch assigns the first
available unused port (if any).
It is possible to connect several cables in the same switch port:
in this way all this cables work concurrently.
It means that packet can result as duplicate but no ARP table loops are
generated.
Is useful when vde is used for mobility. Several physical interfaces can
be used at a time during handoffs to prevent hichups in connectivity.
.TP
\fB\-g\fP, \fB\-\-group\fP=\fIgroup\fP
group ownership of the communication socket. For security when more
want to share a switch it is better to use a unix group to own the comm sockets
so that the network traffic cannot be sniffed.
.TP
\fB\-m\fP, \fB\-\-mod\fP=\fIoctal-mode\fP
octal chmod like permissions for the comm sockets
.TP
\fB\-s\fP, \fB\-\-sock\fP=\fIsocket\fP
specify the UNIX socket to be used by local programs for joining the VDE.
The default value is "/tmp/vde.ctl".
.TP
\fB\-d\fP, \fB\-\-daemon\fP
start vde_plug2tap as a background process
.TP
\fB\-P\fP, \fB\-\-pidfile\fP=\fIpidfile\fP
put the process ID of vde_plug2tap in \fIpidfile\fP. Can be used with
\fB\-\-daemon\fP to store the PID for future killing.
.TP
\fB\-h\fP, \fB\-\-help\fP
show a brief help

.SH NOTICE
Virtual Distributed Ethernet is not related in any way with
www.vde.com ("Verband der Elektrotechnik, Elektronik und Informationstechnik"
i.e. the German "Association for Electrical, Electronic & Information
Technologies").

.SH SEE ALSO
\fBvde_switch\fP(1),
\fBvde_plug\fP(1),
\fBdpipe\fP(1).
.br
.SH AUTHOR
VDE is a project by Renzo Davoli <renzo@cs.unibo.it>


Changelog
VDE 2.3.3, 2022-05-02
	* Several bug fixes
	* new hash function
	* Allow pipes on stdin for daemonized switches
	* fixed wirefilter '-N' option
	* vde_switch: min packet len enforced
	* Added vde_router
	* dpipe: Fixed error checking for setpgrp().
	* Fix incorrect chacha key size
	* configure: do not test for CXX, it's not needed
	* Fix description for "vlan/delport" command.
	* Silence compiler warning about implicit type.
	* Fix typo in error message
	* Fixes 2019.10.06
	* fstp: Add static to inline functions
	* configure: Check for wolfSSL instead of OpenSSL.
	* build: Fix parallel make.
	* vde_switch: add `-nostdin` flag to allow stdin to be closed
	* Deprecate modules
	* Fix warnings with newer gcc
	* [vde_plug] Removed kvde IPN support
	* In vde_switch retry send() when it returns ENOBUFS on Darwin and FreeBSD
	* configure.ac: Downgrade autoconf from 2.71 to 2.59
	* Added GitHub actions

VDE 2.3.2, 2011-11-23
	* dpipe: new daemon mode
	* [vde_switch] new mainloop management
	* [vde_switch] hash table based on 64bits integers
	* [vde_switch] new priority queue for packets
	* wirefilter: capacity renamed as "channel buffer size"

VDE 2.3.1, 2010-07-25
	* [vde_switch] Fix control socket path resolution regression.

VDE 2.3.0, 2010-05-29
	* Remove some hardcoded paths and use SYSCONFDIR/LOCALSTATEDIR instead.
	* [vde_switch] Fix multiple vlan tagging when packetfilter is used.
	* Build system fixes.
	* Ship README and patch to add VDE support in VirtualBox (included in
	  VirtualBox since 3.2.0).
	* Ship an updated version of slirp taken from QEMU and include VDE support.
	* [vde_cryptcab] Remove SIGALRM-based timers.
	* [vde_l3] Fix unicast test (sourceforge: #2726725)
	* [libvdeplug] Add vdestream abstraction to send/receive bytes (previously in
	  vde_plug)
	* [vde_switch] New command "port/createauto" to create a new port with an
	  automatically allocated ID

VDE 2.2.3, 2009-05-05
	* [vdeterm] Moved history management in a separate library.
	* [vde_switch] Console bug fixes.
	* Added vde_pcapplug (connect a vde_switch to a real interface using pcap)
	* Added missing includes of limits.h/ctype.h
	* [vde_switch] Added option --mgmtgroup to consmgmt module to specify the
	  group of the management socket.
	* [vde_switch] Added option --dirmode to datasock module to specify the mode
	  of the socket directory.
	* [vdetaplib] Added specification of port, mode and group.
	* [kvde_switch] Bugfixes.
	* [vde_switch] pdump plugin: add fifo support and buffered/unbuffered choice.
	* [common] Fix '#include <limits.h>' for PATH_MAX. (Closes: #2023790)
	* [vdetaplib] Manpage: added an example to override default port, group and
	  mode. (Closes: #2066885)
	* [vde_switch] Resize untagged ports bitarray while changing number of ports
	  (Thanks to Michael Sallaway). Closes: #2115547.
	* [vde_switch] Fix memory reset in BA_REALLOC and BAC_REALLOC (by rd235).
	  (Closes: #2123601)
	* [kvde_switch] By default kvde_switch is now not compiled in.
	* [vdeq] Fixed loop if an unknow parameter is specified. Thanks to tfero.
	  (Closes: #2134438)
	* [slirpvde] Imported patch from <ludwig.nussel@suse.de> to fix static buffer
	  overflow (Closes: #2138410).
	* [vde_switch] Fixed compatibility with armel architecture using "attribute
	  packet" on registration message structures (thanks to Michele Cucchi).
	* [vde_cryptcap] Some modifications to allow the client to work on the
	  OpenMoko freerunner.
	* [vde_switch] Added port access control.
	* [vde_switch] Changed regular user standard switch to ~/.vde2/stdsock.
	* [libvdeplug] Downsize vde_open_real() and vde_realpath() stack to fit in
	  User-Mode Linux.
	* [common] Add pkg-config files for VDE libraries.
	* [common] Add '--disable-pcap' configure option, thanks to Wulf C. Krueger
	  (Closes: #2600817).
	* [common] Fix bad malloc replacement (Closes: #2631581), thanks to Michele
	  Cucchi.
	* [vde_cryptcab] Added support for daemonization.
	* [vde_cryptcab] Fixed server side disconnections (thanks to Brendan Grieve).
	* [wirefilter] Added rc file.

VDE 2.2.2, 2008-07-08
	* Added --enable-profile to configure (for profiling)
	* Added cleanup() in vdeq on successfull termination (thanks huslu)
	* Added --mod to new syntax in vdeq man page
	* Fixed fd leak in runscript() (Closes: #2009311 on SF)
	* wirefilter man and code cleanup
	* wirefilter option for blinking added.
	* Gaussian distribution for range values, Gilbert bursty model for packet loss.
	* Bugfix on mgmt sessions.
	* Poor's man hub implemented. (patch by Luca Saiu)

VDE 2.2.1, 2008-06-17
	* New contact email address: <info@v2.cs.unibo.it>
	* Code cleanup (some more warnings removed)
	* [vde_crytcab] Cleared large memory leak due to missing
	  EVP_CIPHER_CTX_cleanup() after encryption/decryption
	* [vde_cryptcab] Cleared (very small) memory leak in crc32 calls
	* [vde_cryptcab] Fixed bug on handover after session timeout: now client
	  re-establishes the connection when it sees some traffic, after session
	  timeout.
	* Heavy directories reorganization, all source code moved under src and
	  include
	* New internal library libvdecommon.la with common utilities and LIBOBJ
	  replacement/compatibility functions
	* [common] Fix in poll emulation management
	* Fix to distclean problem due to wrong LIBOBJ management
	* [vde_switch] hash/setsize 0 was not recognized as an error and caused vde_switch
	  to hang due to an infinite loop
	* Don't compile cryptcab if configure detects not to.
	* Fix for autotools in Slackware
	* [vde_cryptcab] refactoring:
	  - simplified protocol (though keeping it back-compatible)
	  - fixed some bugs and memory leaks
	  - added some new features
	* [vde_switch] fix on BPDU generation/handling: do not generate BPDU if FSTP is not active
	  or we are HUB, forward BPDU if FSTP is not active (Closes #1943973 on SF).
	* [slirpvde] include limits.h on linux as well.
	* [vde_switch] Applied patch from Bjorn Mork to remove 32/64 bit confusion from bitarray.h

src/vde_pcapplug.c
/*
 * Copyright (C) 2008 - Luca Bigliardi
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */

#include <stdio.h>
#include <fcntl.h>
#include <stdlib.h>
#include <unistd.h>
#include <signal.h>
#include <syslog.h>
#include <errno.h>
#include <string.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/socket.h>
#include <stdarg.h>
#include <syslog.h>
#include <libgen.h>
#define _GNU_SOURCE
#include <getopt.h>
#include <pcap.h>
#include <limits.h>

#include <config.h>
#include <vde.h>
#include <vdecommon.h>
#include <libvdeplug.h>

#ifdef VDE_FREEBSD
#include <sys/socket.h>
#endif

#if defined(VDE_DARWIN) || defined(VDE_FREEBSD)
#	if defined HAVE_SYSLIMITS_H
#		include <syslimits.h>
#	elif defined HAVE_SYS_SYSLIMITS_H
#		include <sys/syslimits.h>
#	else
#		error "No syslimits.h found"
#	endif
#endif

#define BUFSIZE 2048

static VDECONN *conn = NULL;
static pcap_t *pcap = NULL;

char *prog;
int logok;
static char *pidfile = NULL;
static char pidfile_path[PATH_MAX];

void printlog(int priority, const char *format, ...)
{
	va_list arg;

	va_start (arg, format);

	if (logok)
		vsyslog(priority,format,arg);
	else {
		fprintf(stderr,"%s: ",prog);
		vfprintf(stderr,format,arg);
		fprintf(stderr,"\n");
	}
	va_end (arg);
}

static void cleanup(void)
{
	if((pidfile != NULL) && unlink(pidfile_path) < 0) {
		printlog(LOG_WARNING,"Couldn't remove pidfile '%s': %s", pidfile, strerror(errno));
	}

	if (pcap)
		pcap_close(pcap);
	if (conn)
		vde_close(conn);
}

static void sig_handler(int sig)
{
	cleanup();
	signal(sig, SIG_DFL);
	if (sig == SIGTERM)
		_exit(0);
	else
		kill(getpid(), sig);
}

static void setsighandlers()
{
	/* setting signal handlers.
	 * sets clean termination for SIGHUP, SIGINT and SIGTERM, and simply
	 * ignores all the others signals which could cause termination. */
	struct { int sig; const char *name; int ignore; } signals[] = {
		{ SIGHUP, "SIGHUP", 0 },
		{ SIGINT, "SIGINT", 0 },
		{ SIGPIPE, "SIGPIPE", 1 },
		{ SIGALRM, "SIGALRM", 1 },
		{ SIGTERM, "SIGTERM", 0 },
		{ SIGUSR1, "SIGUSR1", 1 },
		{ SIGUSR2, "SIGUSR2", 1 },
		{ SIGPROF, "SIGPROF", 1 },
		{ SIGVTALRM, "SIGVTALRM", 1 },
#ifdef VDE_LINUX
		{ SIGPOLL, "SIGPOLL", 1 },
#ifdef SIGSTKFLT
		{ SIGSTKFLT, "SIGSTKFLT", 1 },
#endif
		{ SIGIO, "SIGIO", 1 },
		{ SIGPWR, "SIGPWR", 1 },
#ifdef SIGUNUSED
		{ SIGUNUSED, "SIGUNUSED", 1 },
#endif
#endif
#ifdef VDE_DARWIN
		{ SIGXCPU, "SIGXCPU", 1 },
		{ SIGXFSZ, "SIGXFSZ", 1 },
#endif
		{ 0, NULL, 0 }
	};

	int i;
	for(i = 0; signals[i].sig != 0; i++)
		if(signal(signals[i].sig,
					signals[i].ignore ? SIG_IGN : sig_handler) < 0)
			perror("Setting handler");
}

struct pollfd pollv[]={{0,POLLIN|POLLHUP},{0,POLLIN|POLLHUP},{0,POLLIN|POLLHUP}};

static void usage(void) {
	fprintf(stderr, "Usage: %s [OPTION]... interface\n\n", prog);
	fprintf(stderr, "  -p, --port=portnum          Port number in the VDE switch\n"
			        "  -g, --group=group           Group for the socket\n"
					"  -m, --mode=mode             Octal mode for the socket\n"
					"  -s, --sock=socket           VDE switch control socket or dir\n"
					"  -d, --daemon                Launch in background\n"
					"  -P, --pidfile=pidfile       Create pidfile with our PID\n"
					"  -h, --help                  This help\n");
	exit(-1);
}

unsigned char bufin[BUFSIZE];

static void save_pidfile()
{
	if(pidfile[0] != '/')
		strncat(pidfile_path, pidfile, PATH_MAX - strlen(pidfile_path) - 1);
	else
		strncpy(pidfile_path, pidfile, PATH_MAX - 1);

	int fd = open(pidfile_path,
			O_WRONLY | O_CREAT | O_EXCL,
			S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
	FILE *f;

	if(fd == -1) {
		printlog(LOG_ERR, "Error in pidfile creation: %s", strerror(errno));
		exit(1);
	}

	if((f = fdopen(fd, "w")) == NULL) {
		printlog(LOG_ERR, "Error in FILE* construction: %s", strerror(errno));
		exit(1);
	}

	if(fprintf(f, "%ld\n", (long int)getpid()) <= 0) {
		printlog(LOG_ERR, "Error in writing pidfile");
		exit(1);
	}

	fclose(f);
}

void pcap_callback(u_char *u, const struct pcap_pkthdr *h, const u_char *data)
{
	vde_send(conn, data, h->len, 0);
}

void setup_fd(int fd)
{
/* FreeBSD settings */
#if defined(VDE_FREEBSD)
	/*
	 * Tell the kernel that the header is fully-formed when it gets it.
	 * This is required in order to fake the src address.
	 */
	{ unsigned int i = 1; ioctl(fd, BIOCSHDRCMPLT, &i); }
	/*                    
	 * Tell the kernel that the packet has to be processed immediately.
	 */
	{ unsigned int i = 1; ioctl(fd, BIOCIMMEDIATE, &i); }
	/*
	 * Allow guest-host communication.
	 */
	{ unsigned int i = 1; ioctl(fd, BIOCFEEDBACK, &i); }
#endif
/* 
 * BIG TODO(shammash):
 * let host and guest communicate under linux
 */
/*
 * Most important parts of libpcap with PF_PACKET on Linux:
	rawfd = socket(PF_PACKET, SOCK_RAW, htons(ETH_P_ALL));
	iface_get_id(int fd, const char *device, char *ebuf)
	{
		struct ifreq    ifr;

		memset(&ifr, 0, sizeof(ifr));
		strncpy(ifr.ifr_name, device, sizeof(ifr.ifr_name));

		if (ioctl(fd, SIOCGIFINDEX, &ifr) == -1) {
			snprintf(ebuf, PCAP_ERRBUF_SIZE,
					"SIOCGIFINDEX: %s", pcap_strerror(errno));
			return -1;
		}       

		return ifr.ifr_ifindex;
	}

	struct packet_mreq mr;
	memset(&mr, 0, sizeof(mr));
        mr.mr_ifindex = handle->md.ifindex;
        mr.mr_type    = PACKET_MR_PROMISC;
        if (setsockopt(sock_fd, SOL_PACKET, PACKET_ADD_MEMBERSHIP, &mr, sizeof(mr)) == -1) {
		snprintf(ebuf, PCAP_ERRBUF_SIZE, "setsockopt: %s", pcap_strerror(errno));
        }
*
*/
#if defined(VDE_LINUX)
	{
		unsigned int i = 1;
		if (setsockopt(fd, SOL_SOCKET, SO_BROADCAST, &i, sizeof(i)) == -1) {
			printlog(LOG_ERR, "SO_BROADCAST: %s\n", strerror(errno));
			exit(1);
		}
	}
#endif
}

int main(int argc, char **argv)
{
	static char *sockname=NULL;
	static char *ifname=NULL;
	int daemonize=0;
	char errbuf[PCAP_ERRBUF_SIZE];
	int pcapfd;
	register ssize_t nx;
	struct vde_open_args open_args={.port=0,.group=NULL,.mode=0700};
	int c;
	prog=argv[0];
	while (1) {
		int option_index = 0;

		static struct option long_options[] = {
			{"sock", 1, 0, 's'},
			{"port", 1, 0, 'p'},
			{"help",0,0,'h'},
			{"mod",1,0,'m'},
			{"group",1,0,'g'},
			{"daemon",0,0,'d'},
			{"pidfile", 1, 0, 'P'},
			{0, 0, 0, 0}
		};
		c = GETOPT_LONG (argc, argv, "hdP:p:s:m:g:",
				long_options, &option_index);
		if (c == -1)
			break;

		switch (c) {
			case 'p':
				open_args.port=atoi(optarg);
				if (open_args.port <= 0)
					usage(); //implies exit
				break;

			case 'h':
				usage(); //implies exit
				break;

			case 's':
				sockname=strdup(optarg);
				break;

			case 'm':
				sscanf(optarg,"%o",(unsigned int *)&(open_args.mode));
				break;

			case 'g':
				open_args.group=strdup(optarg);
				break;

			case 'd':
				daemonize=1;
				break;

			case 'P':
				pidfile=strdup(optarg);
				break;

			default:
				usage(); //implies exit
		}
	}

	if (daemonize) {
		openlog(basename(prog), LOG_PID, 0);
		logok=1;
		syslog(LOG_INFO,"VDE_PCAPPLUG started");
	}
	/* saves current path in pidfile_path, because otherwise with daemonize() we
	 * forget it */
	if(getcwd(pidfile_path, PATH_MAX-2) == NULL) {
		printlog(LOG_ERR, "getcwd: %s", strerror(errno));
		exit(1);
	}
	strcat(pidfile_path, "/");

	conn=vde_open(sockname,"vde_pcapplug:",&open_args);
	if (conn == NULL) {
		printlog(LOG_ERR,"vde_open %s: %s",sockname?sockname:"DEF_SWITCH",strerror(errno));
		exit(1);
	}

	if (daemonize && daemon(0, 0)) {
		printlog(LOG_ERR,"daemon: %s",strerror(errno));
		exit(1);
	}

	/* once here, we're sure we're the true process which will continue as a
	 * server: save PID file if needed */
	if(pidfile) save_pidfile();

	if (optind < argc)
		ifname=argv[optind];
	else
		usage(); // implies exit
	
	atexit(cleanup);
	setsighandlers();

	pcap = pcap_open_live(ifname, BUFSIZE, 1, 0, errbuf);
	if (pcap == NULL) {
		printlog(LOG_ERR, "Open %s: %s\n", ifname, errbuf);
		exit(1);
	}
	if (pcap_datalink(pcap) != DLT_EN10MB ) {
		printlog(LOG_ERR, "Given interface is not ethernet\n");
		exit(1);
	}
	pcapfd=pcap_get_selectable_fd(pcap);
	if (pcapfd == -1) {
		printlog(LOG_ERR, "pcap has no fd for poll()\n");
		exit(1);
	}
	setup_fd(pcapfd);

	pollv[0].fd=pcapfd;
	pollv[1].fd=vde_datafd(conn);
	pollv[2].fd=vde_ctlfd(conn);

	for(;;) {
		poll(pollv,3,-1);
		if ((pollv[0].revents | pollv[1].revents | pollv[2].revents) & POLLHUP ||
				pollv[2].revents & POLLIN) 
			break;
		if (pollv[0].revents & POLLIN) {
			nx = pcap_dispatch(pcap, 1, &pcap_callback, NULL);
			if (nx<=0)
				break;
		}
		if (pollv[1].revents & POLLIN) {
			nx=vde_recv(conn,bufin,sizeof(bufin),0);
			if (nx<=0)
				break;
			nx = pcap_inject(pcap, bufin, nx);
			if (nx<=0)
				break;
		}
	}
	return(0);
}
